javascript/node/selenium-webdriver/bidi/bidi.js

```diff
@@ -0,0 +1,217 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+const { EventEmitter } = require('node:events')
+const WebSocket = require('ws')
+
+const RESPONSE_TIMEOUT = 1000 * 30
+
+class BIDI extends EventEmitter {
+  id = 0
+  isConnected = false
+  events = []
+  browsingContexts = []
+
+  /**
+   * Create a new websocket connection
+   * @param _webSocketUrl
+   */
+  constructor (_webSocketUrl) {
+    super()
+    this._ws = new WebSocket(_webSocketUrl)
+  }
+
+  /**
+   * open connection to fire an open event
+   * @returns {Promise<resolved>}
+   */
+  connect () {
+    return new Promise((resolve) => this._ws.on('open', () => {
+      this.isConnected = true
+      resolve()
+    }))
+  }
+
+  /**
+   * @returns {WebSocket}
+   */
+  get socket () {
+    return this._ws
+  }
+
+  /**
+   * @returns {boolean|*}
+   */
+  get isConnected () {
+    return this.isConnected
+  }
+
+  /**
+   * Sends a bidi request
+   * @param params
+   * @returns {Promise<unknown>}
+   */
+  send (params) {
+    if (!this.isConnected) {
+      throw new Error('No connection to WebDriver Bidi was established')
+    }
+
+    const id = ++this.id
+
+    this._ws.send(JSON.stringify({ id, ...params }))
+
+    return new Promise((resolve, reject) => {
+      const t = setTimeout(() => {
```

Comment:
Not a big fan of single character variable names. We don't pay per character

Reply:
Apologies, my bad! 
Updated!